### DISCUSSION TITLE
NTP accuracy - how is it possible?

### DISCUSSION THREAD

**Discussion by davepl** [PlummersSoftwareLLC/NightDriverStrip#377] [General]
Right now I've got NTP checking the time every 30 seconds, just temporarily.  And I noticed that the time coming back from the server is ZERO delta from the time on the ESP32 once set.

The code plainly pulls second and MICROSECONDS from the packet, and yet they match the local clock down to the microsecond ever time.

Is it really possible that both NTP and the ESP32's RTC are accurate within one millionth of a second over a 30 second span?

The NTP protocol must be pretty amazing at what it does for this to work as well!

```
 // Obtain the time from the packet, convert to Unix time, and adjust for the time zone.

    struct timeval tvNew = { 0 };

    uint32_t frac  = (uint32_t) chNtpPacket[44] << 24
                    | (uint32_t) chNtpPacket[45] << 16
                    | (uint32_t) chNtpPacket[46] <<  8
                    | (uint32_t) chNtpPacket[47] <<  0;

    uint32_t microsecs = ((uint64_t) frac * 1000000) >> 32;

    // BUGBUG (davepl): I've been gettin back odd packets where the clock is year 2036 and micros is 0. I ignore those.

    if (microsecs == 0)
    {
        debugW("Bogus NTP time received, ignoring.");
        return false;
    }

    debugV("NTP clock: Raw values sec=%u, usec=%u", frac, microsecs);

    tvNew.tv_sec = ((unsigned long)chNtpPacket[40] << 24) +       // bits 24 through 31 of ntp time
        ((unsigned long)chNtpPacket[41] << 16) +                        // bits 16 through 23 of ntp time
        ((unsigned long)chNtpPacket[42] << 8) +                         // bits  8 through 15 of ntp time
        ((unsigned long)chNtpPacket[43]) -                              // bits  0 through  7 of ntp time
        (((70UL * 365UL) + 17UL) * 86400UL);                            // ntp to unix conversion

    tvNew.tv_usec = microsecs;

    timeval tvOld;
    gettimeofday(&tvOld, nullptr);

    double dOld = tvOld.tv_sec + (tvOld.tv_usec / (float) MICROS_PER_SECOND);
    double dNew = tvNew.tv_sec + (tvNew.tv_usec / (float) MICROS_PER_SECOND);
```
When this code runs, dOld == dNew.  They're the same.  Every time except the first time!

**Comment by robertlipe** [comment]
There may be a lower level that's running an NTP daemon behind your back. IIRC, PlatformIO sits atop Arduino framework which is atop ESP-IDF. ESP-IDF used LWIP for networking and LWIP can be configured to run the equivalent of ntpd. That call on boot to settimeofday() may have kicked that off for you.

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/system_time.html?highlight=rtc%20accuracy

If that's indeed configured in and turned on in this build, the projects NDP implementation may be moot. Get the time on system startup or when a user changes it and maybe you're done. I don't think this project cares much about low power modes and going to sleep in one time zone but waking up in another time zone after the DST changeover the day before and such.

The Espressif IDF is a pretty fully features "OS" unto itself.

**Comment by davepl** [comment]
It was a bug!  A float was being added to a double, but the scale of the integer portion was so big that the decimal part was being lost in the float conversion.

Fixed in v40
